caddyhttp/basicauth/basicauth.go

```diff
@@ -63,7 +63,7 @@ func (a BasicAuth) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error
 
 			// let upstream middleware (e.g. fastcgi and cgi) know about authenticated
 			// user; this replaces the request with a wrapped instance
-			r = r.WithContext(context.WithValue(r.Context(),
+			*r = *r.WithContext(context.WithValue(r.Context(),
```

Comment:
I should let Toby answer, but I wanted to better understand myself, so pardon me @tobya  😬 

@mholt  I believe its because "under the hood" it's still [passing by value](https://groups.google.com/forum/m/#!topic/golang-nuts/INedfATw74A) [Rob Pike] and not by reference.   

`WithContext` is [returning  a new `http.Request` (r2)](https://golang.org/src/net/http/request.go?s=11679:11738#L317).   So we need to dereference to modify the original `http.Request` that r is pointing to,  otherwise you get an entirely new `r` and the new context key doesn't get passed down the chain.   My own testing shows this to be the case.

I have output dumps of this if you need to see it.    My research shows the only way to avoid this is to go all the way back up the chain to where the first ServeHTTP is called,  or just do the dereference within the middleware